crates/sui-framework/sources/linked_table.move

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Similar to `sui::table` but the values are linked together. Allowing for ordered insertion and
+/// removal
+module sui::linked_table {
+
+use std::option::{Self, Option};
+use sui::object::{Self, UID};
+use sui::dynamic_field as field;
+use sui::tx_context::TxContext;
+
+// Attemped to destroy a non-empty table
+const ETableNotEmpty: u64 = 0;
+// Attemped to remove the front or back of an empty table
+const ETableIsEmpty: u64 = 1;
+
+struct LinkedTable<K: copy + drop + store, phantom V: store> has key, store {
+    /// the ID of this table
+    id: UID,
+    /// the number of key-value pairs in the table
+    size: u64,
+    /// the front of the table, i.e. the key of the first entry
+    head: Option<K>,
```

Comment:
IIUC, `head` is only `None` when `tail` is also `None`. If that's right, inserting an intermediate type like

```
struct Bookends {
  head: K,
  tail: K
}
```

and using `Option<Bookends>` here might make sense, since will it enforce that invariant by construction.

I think it will simplify the handling of `head/tail` in a few places too.